include/oscore/oscore_context.h

```diff
@@ -0,0 +1,271 @@
+  HASH_ADD(hh, (r), token->s[0], (obj)->token->length, (obj))
+
+#define OSCORE_ASSOCIATIONS_DELETE(r, obj) HASH_DELETE(hh, (r), (obj))
+
+#define OSCORE_ASSOCIATIONS_ITER(r, tmp)                                       \
+  oscore_associations_t *tmp, *rtmp;                                           \
+  HASH_ITER (hh, (r), tmp, rtmp)
+
+#define OSCORE_ASSOCIATIONS_ITER_SAFE(e, el, rtmp)                             \
+  for ((el) = (e); (el) && ((rtmp) = (el)->hh.next, 1); (el) = (rtmp))
+
+#define OSCORE_ASSOCIATIONS_FIND(r, k, res)                                    \
+  { HASH_FIND(hh, (r), (k)->s, (k)->length, (res)); }
+
+struct oscore_association_t {
+  UT_hash_handle hh;
+  oscore_recipient_ctx_t *recipient_ctx;
+  coap_pdu_t *sent_pdu;
+  coap_bin_const_t *token;
+  coap_bin_const_t *aad;
+  coap_bin_const_t *nonce;
+  coap_bin_const_t *partial_iv;
+  coap_tick_t last_seen;
+  uint8_t is_observe;
+};
+
+/**
+ * oscore_derive_ctx - derive a osc_ctx fro oscore_conf information
+ *
+ * @param c_context The CoAP context to associate OSCORE context with.
+ * @param oscore_conf The OSCORE configuration to use.
+ *
+ * @return NULL if failure or derived OSCORE context linked onto
+ *         @p c_context chain.
+ */
+oscore_ctx_t *oscore_derive_ctx(coap_context_t *c_context,
+                                coap_oscore_conf_t *oscore_conf);
+
+/**
+ * oscore_duplicate_ctx - duplicate a osc_ctx
+ *
+ * @param c_context The CoAP context to associate OSCORE context with.
+ * @param o_osc_ctx The OSCORE context to duplicate.
+ * @param sender_id The Sender ID to use in the duplication.
+ * @param recipient_id The Recipient ID to use in the duplication.
+ * @param id_context The Context ID to use in the duplication.
+ *
+ * @return NULL if failure or duplicated OSCORE context linked onto
+ *         @p c_context chain.
+ */
+oscore_ctx_t *oscore_duplicate_ctx(coap_context_t *c_context,
+                                   oscore_ctx_t *o_osc_ctx,
+                                   coap_bin_const_t *sender_id,
+                                   coap_bin_const_t *recipient_id,
+                                   coap_bin_const_t *id_context);
+
+/**
+ * oscore_update_ctx - update a osc_ctx with a new id_context
+ *
+ * @param osc_ctx The OSCORE context to update.
+ * @param id_context The Context ID to use in the duplication.
+ */
+void oscore_update_ctx(oscore_ctx_t *osc_ctx, coap_bin_const_t *id_context);
+
+void oscore_free_context(oscore_ctx_t *osc_ctx);
+
+void oscore_free_contexts(coap_context_t *c_context);
+
+int oscore_remove_context(coap_context_t *c_context, oscore_ctx_t *osc_ctx);
+
+oscore_recipient_ctx_t *oscore_add_recipient(oscore_ctx_t *ctx,
+                                             coap_bin_const_t *rid,
+                                             uint32_t break_key);
+
+int oscore_delete_recipient(oscore_ctx_t *osc_ctx, coap_bin_const_t *rid);
+
+int _strcmp(const char *a, const char *b);
```

Comment:
This function is never called, so I suggest removing it.

Reply:
Removed.